File: server/src/metadata/hentainexus.rs
use super::MultiTextField;
use crate::{db, utils};
use serde::Deserialize;
use slug::slugify;

#[derive(Deserialize)]
pub struct Metadata {
  #[serde(rename = "Title")]
  pub title: String,
  #[serde(rename = "Description", default)]
  pub description: Option<String>,
  #[serde(rename = "Source", default)]
  pub source: Option<String>,
  #[serde(rename = "URL", default)]
  pub url: Option<String>,
  #[serde(rename = "Artist", default)]
  pub artists: Option<MultiTextField>,
  #[serde(rename = "Circle", default)]
  pub circles: Option<MultiTextField>,
  #[serde(rename = "Magazine", default)]
  pub magazines: Option<MultiTextField>,
  #[serde(rename = "Event", default)]
  pub events: Option<MultiTextField>,
  #[serde(rename = "Publisher", default)]
  pub publishers: Option<MultiTextField>,
  #[serde(rename = "Parody", default)]
  pub parodies: Option<MultiTextField>,
  #[serde(rename = "Tags", default)]
  pub tags: Option<Vec<String>>,
  #[serde(rename = "Thumbnail", default)]
  pub thumb_index: Option<i16>,
  #[serde(rename = "Released", default)]
  pub released: Option<i64>,
}

pub fn add_metadata(info: Metadata, archive: &mut db::UpsertArchiveData) -> anyhow::Result<()> {
  archive.title = Some(info.title);
  archive.slug = archive.title.as_ref().map(slugify);
  archive.description = info.description;
  archive.thumbnail = info.thumb_index.map(|index| index + 1);
  archive.released_at = utils::map_timestamp(info.released);

  archive.artists = info.artists.map(|field| field.to_vec());
  archive.circles = info.circles.map(|field| field.to_vec());
  archive.magazines = info.magazines.map(|field| field.to_vec());
  archive.events = info.events.map(|field| field.to_vec());
  archive.publishers = info.publishers.map(|field| field.to_vec());
  archive.parodies = info.parodies.map(|field| field.to_vec());
  archive.tags = info.tags.map(|tags| {
    tags
      .into_iter()
      .map(|tag| (utils::capitalize_words(&tag), "".to_string()))
      .collect()
  });

  let mut sources: Vec<db::ArchiveSource> = vec![];

  if let Some(url) = info.url {
    sources.push(db::ArchiveSource {
      name: utils::parse_source_name(&url),
      url: Some(url),
    });
  }

  if let Some(url) = info.source {
    sources.push(db::ArchiveSource {
      name: utils::parse_source_name(&url),
      url: Some(url),
    });
  }

  if !sources.is_empty() {
    archive.sources = Some(sources);
  }

  Ok(())
}

File: cli/metadata/hentainexus.ts
import { type ArchiveMetadata } from '../../shared/metadata';
import { mapMultiField, multiTextField } from './schemas';
import dayjs from 'dayjs';
import YAML from 'yaml';
import { z } from 'zod';

const metadataSchema = z.object({
	Title: z.string(),
	Description: z.string().optional(),
	Source: z.string().optional(),
	URL: z.string().optional(),
	Artist: multiTextField.optional(),
	Circle: multiTextField.optional(),
	Magazine: multiTextField.optional(),
	Event: multiTextField.optional(),
	Publisher: multiTextField.optional(),
	Parody: multiTextField.optional(),
	Tags: z.array(z.string()).optional(),
	Thumbnail: z.number().optional(),
	Released: z.number().optional(),
});

export default async (content: string, archive: ArchiveMetadata) => {
	const parsed = YAML.parse(content);
	const { data, error } = metadataSchema.safeParse(parsed);

	if (!data) {
		throw new Error(`Failed to parse HentaiNexus metadata: ${error}`);
	}

	archive = structuredClone(archive);

	archive.title = data.Title;
	archive.description = data.Description;
	archive.thumbnail = data.Thumbnail !== undefined ? data.Thumbnail + 1 : undefined;
	archive.releasedAt = data.Released ? dayjs.unix(data.Released).toDate() : undefined;

	archive.tags = [];

	archive.tags.push(
		...mapMultiField(data.Artist).map((tag) => ({ namespace: 'artist', name: tag }))
	);
	archive.tags.push(
		...mapMultiField(data.Circle).map((tag) => ({ namespace: 'circle', name: tag }))
	);
	archive.tags.push(
		...mapMultiField(data.Magazine).map((tag) => ({ namespace: 'magazine', name: tag }))
	);
	archive.tags.push(...mapMultiField(data.Event).map((tag) => ({ namespace: 'event', name: tag })));
	archive.tags.push(
		...mapMultiField(data.Publisher).map((tag) => ({ namespace: 'publisher', name: tag }))
	);
	archive.tags.push(
		...mapMultiField(data.Parody).map((tag) => ({ namespace: 'parody', name: tag }))
	);

	if (data.Tags) {
		archive.tags.push(...data.Tags.map((tag) => ({ namespace: 'tag', name: tag })));
	}

	archive.sources = [];

	if (data.URL) {
		archive.sources.push({ url: data.URL });
	}

	if (data.Source) {
		archive.sources.push({ url: data.Source });
	}

	return archive;
};
